File: Send-Email-SMS-Azure-AppService-MessageFlow-python/sms_service.py
import json
import requests
import logging
from typing import List, Optional, Dict, Any
from dataclasses import dataclass, asdict

@dataclass
class SmsRequest:
    sender: str
    message: str
    phone_numbers: List[str]
    validity: Optional[int] = None
    schedule_time: Optional[int] = None
    type: Optional[int] = None
    short_link: Optional[bool] = None
    webhook_url: Optional[str] = None
    external_id: Optional[str] = None

@dataclass
class SmsResponse:
    success: bool
    message: str
    response_content: str = ""

class SmsService:
    def __init__(self, config: Dict[str, Any]):
        """
        Initialize SMS service with configuration
        
        Args:
            config: Dictionary containing 'rest_api' configuration with 'authorization' and 'application_key'
        """
        self.config = config
        self.logger = logging.getLogger(__name__)
        
        # Validate configuration
        if not config.get('rest_api', {}).get('authorization'):
            raise ValueError("REST API authorization is missing in configuration")
        if not config.get('rest_api', {}).get('application_key'):
            raise ValueError("REST API application key is missing in configuration")
    
    def send_sms(self, request: SmsRequest) -> SmsResponse:
        """
        Send SMS using MessageFlow API
        
        Args:
            request: SmsRequest object containing SMS details
            
        Returns:
            SmsResponse object with success status and message
        """
        self.logger.info(f"SmsService.send_sms called with request: {json.dumps(asdict(request), indent=2)}")
        
        try:
            # Prepare headers
            headers = {
                'Authorization': self.config['rest_api']['authorization'],
                'Application-Key': self.config['rest_api']['application_key'],
                'Content-Type': 'application/json'
            }
            
            self.logger.info(f"Headers set - Authorization: {self.config['rest_api']['authorization'][:10]}..., Application-Key: {self.config['rest_api']['application_key'][:10]}...")
            
            # Prepare payload
            payload = {
                'sender': request.sender,
                'message': request.message,
                'phoneNumbers': request.phone_numbers,
                'validity': request.validity,
                'scheduleTime': request.schedule_time,
                'type': request.type,
                'shortLink': request.short_link,
                'webhookUrl': request.webhook_url,
                'externalId': request.external_id
            }
            
            # Remove None values
            payload = {k: v for k, v in payload.items() if v is not None}
            
            json_payload = json.dumps(payload, indent=2)
            self.logger.info(f"SMS API Request Payload: {json_payload}")
            
            # Send request
            self.logger.info("Sending HTTP request to: https://api.messageflow.com/v2.1/sms")
            
            response = requests.post(
                'https://api.messageflow.com/v2.1/sms',
                headers=headers,
                json=payload,
                timeout=30
            )
            
            response_content = response.text
            self.logger.info(f"SMS API Response Status: {response.status_code}")
            self.logger.info(f"SMS API Response Content: {response_content}")
            
            # Parse response
            if response.status_code == 200:
                try:
                    response_data = response.json()
                    if isinstance(response_data, dict):
                        return SmsResponse(
                            success=True,
                            message=response_data.get('message', 'SMS sent successfully'),
                            response_content=response_content
                        )
                except (json.JSONDecodeError, KeyError) as e:
                    self.logger.warning(f"Can not parse response to SmsResponse model, but HTTP status is success: {e}")
                
                return SmsResponse(
                    success=True,
                    message="SMS sent successfully",
                    response_content=response_content
                )
            else:
                return SmsResponse(
                    success=False,
                    message=f"HTTP Error: {response.status_code}",
                    response_content=response_content
                )
                
        except requests.exceptions.RequestException as ex:
            self.logger.error(f"HTTP request failed: {ex}")
            return SmsResponse(
                success=False,
                message=f"HTTP request failed: {str(ex)}",
                response_content=str(ex)
            )
        except Exception as ex:
            self.logger.error(f"Error sending SMS: {ex}")
            return SmsResponse(
                success=False,
                message=f"Error: {str(ex)}",
                response_content=str(ex)
            ) 

File: Send-Email-SMS-Azure-AppService-MessageFlow-NodeJS/MessageFlowSmsEmailApp/services/smsService.js
const axios = require('axios');

class SmsRequest {
    constructor ({
        sender,
        message,
        phoneNumbers,
        validity = null,
        scheduleTime = null,
        type = null,
        shortLink = null,
        webhookUrl = null,
        externalId = null
    }) {
        this.sender = sender;
        this.message = message;
        this.phoneNumbers = phoneNumbers;
        this.validity = validity;
        this.scheduleTime = scheduleTime;
        this.type = type;
        this.shortLink = shortLink;
        this.webhookUrl = webhookUrl;
        this.externalId = externalId;
    }
}

class SmsResponse {
    constructor (success, message, responseContent = '') {
        this.success = success;
        this.message = message;
        this.responseContent = responseContent;
    }
}

class SmsService {
    constructor (config) {
        if (!config?.rest_api?.authorization) {
            throw new Error('REST API authorization is missing in configuration');
        }
        if (!config?.rest_api?.application_key) {
            throw new Error('REST API application key is missing in configuration');
        }

        this.config = config;
        this.logger = console;
    }

    async sendSms(request) {
        this.logger.info(`SmsService.sendSms called with request: ${JSON.stringify(request, null, 2)}`);

        try {
            // Prepare headers
            const headers = {
                'Authorization': this.config.rest_api.authorization,
                'Application-Key': this.config.rest_api.application_key,
                'Content-Type': 'application/json'
            };

            this.logger.info(`Headers set - Authorization: ${this.config.rest_api.authorization.substring(0, 10)}..., Application-Key: ${this.config.rest_api.application_key.substring(0, 10)}...`);

            // Prepare payload
            const payload = {
                sender: request.sender,
                message: request.message,
                phoneNumbers: request.phoneNumbers,
                validity: request.validity,
                scheduleTime: request.scheduleTime,
                type: request.type,
                shortLink: request.shortLink,
                webhookUrl: request.webhookUrl,
                externalId: request.externalId
            };

            // Remove null/undefined values
            Object.keys(payload).forEach(key => {
                if (payload[key] === null || payload[key] === undefined) {
                    delete payload[key];
                }
            });

            this.logger.info(`SMS API Request Payload: ${JSON.stringify(payload, null, 2)}`);
            this.logger.info('Sending HTTP request to: https://api.messageflow.com/v2.1/sms');

            // Send request
            const response = await axios.post(
                'https://api.messageflow.com/v2.1/sms',
                payload,
                {
                    headers,
                    timeout: 30000
                }
            );

            const responseContent = JSON.stringify(response.data);

            this.logger.info(`SMS API Response Status: ${response.status}`);
            this.logger.info(`SMS API Response Content: ${responseContent}`);

            // Parse response
            if (response.status === 200) {
                try {
                    if (typeof response.data === 'object') {
                        return new SmsResponse(
                            true,
                            response.data.message || 'SMS sent successfully',
                            responseContent
                        );
                    }
                } catch (error) {
                    this.logger.warn(`Failed to parse response to SmsResponse model, but HTTP status is success: ${error.message}`);
                }

                return new SmsResponse(
                    true,
                    'SMS sent successfully',
                    responseContent
                );
            } else {
                return new SmsResponse(
                    false,
                    `HTTP Error: ${response.status}`,
                    responseContent
                );
            }

        } catch (error) {
            if (error.response) {
                // HTTP error with response
                this.logger.error(`HTTP request failed: ${error.message}`);

                return new SmsResponse(
                    false,
                    `HTTP Error: ${error.response.status}`,
                    JSON.stringify(error.response.data)
                );
            } else if (error.request) {
                // Request made but no response
                this.logger.error(`HTTP request failed: No response received - ${error.message}`);
                return new SmsResponse(
                    false,
                    `HTTP request failed: ${error.message}`,
                    error.message
                );
            } else {
                // Error setting up request
                this.logger.error(`Error sending SMS: ${error.message}`);
                return new SmsResponse(
                    false,
                    `Error: ${error.message}`,
                    error.message
                );
            }
        }
    }
}

module.exports = {
    SmsService,
    SmsRequest,
    SmsResponse
};
